File: tools/clod-poc/src/source_mesh.ts
// LOD0 page source — built by WELDING same-resolution chunk meshes, never re-extracted.
// Plan §3.1 / §11.2 / invariant I2.

import { PageMesh, PageFootprint, ClodBuildError } from "./types.js";
import { ClodPagesConfig } from "./config.js";
import { meshChunk, WorldBounds } from "./terrain.js";
import { weldVertices, WeldReport } from "./weld.js";

export interface PageSource {
  mesh: PageMesh;
  footprint: PageFootprint;
  weld: WeldReport;
}

/** Concatenate several PageMeshes into one buffer (no welding yet). */
export function concat(meshes: PageMesh[]): PageMesh {
  let nv = 0, ni = 0;
  for (const m of meshes) {
    nv += m.positions.length / 3;
    ni += m.indices.length;
  }
  const positions = new Float32Array(nv * 3);
  const normals = new Float32Array(nv * 3);
  const materials = new Float32Array(nv * 4);
  const indices = new Uint32Array(ni);
  let vOff = 0, iOff = 0;
  for (const m of meshes) {
    positions.set(m.positions, vOff * 3);
    normals.set(m.normals, vOff * 3);
    materials.set(m.materials, vOff * 4);
    for (let i = 0; i < m.indices.length; i++) indices[iOff + i] = m.indices[i] + vOff;
    vOff += m.positions.length / 3;
    iOff += m.indices.length;
  }
  return { positions, normals, materials, indices };
}

/**
 * Build a LOD0 page source from its PxP chunks (page coords pageX,pageZ).
 * Step order mirrors §11.2: require PxP chunks -> concat (origins already applied in
 * world space by meshChunk) -> weld internal chunk borders -> outer border preserved.
 */
export function buildLod0PageSource(
  pageX: number,
  pageZ: number,
  cfg: ClodPagesConfig,
  world: WorldBounds,
): PageSource {
  const P = cfg.page.chunks_per_page;
  const S = cfg.page.chunk_size;

  const chunks: PageMesh[] = [];
  for (let dz = 0; dz < P; dz++) {
    for (let dx = 0; dx < P; dx++) {
      chunks.push(meshChunk(pageX * P + dx, pageZ * P + dz, cfg, world));
    }
  }
  if (chunks.length !== P * P) {
    throw new ClodBuildError("PageIncomplete", `expected ${P * P} chunks, got ${chunks.length}`);
  }

  const merged = concat(chunks);
  const { mesh, report } = weldVertices(merged, cfg.simplify.weld_epsilon_cells);

  const footprint: PageFootprint = {
    minX: pageX * P * S,
    minZ: pageZ * P * S,
    maxX: (pageX + 1) * P * S,
    maxZ: (pageZ + 1) * P * S,
  };

  return { mesh, footprint, weld: report };
}

File: tools/clod-rs/src/source_mesh.rs
//! LOD0 page source — welded chunk meshes, never re-extracted (§11.2 / invariant I2).

use crate::config::ClodPagesConfig;
use crate::terrain::{mesh_chunk, WorldBounds};
use crate::types::{ClodBuildError, PageFootprint, PageMesh};
use crate::weld::{weld_vertices, WeldReport};

pub struct PageSource {
    pub mesh: PageMesh,
    pub footprint: PageFootprint,
    pub weld: WeldReport,
}

/// Concatenate several PageMeshes into one buffer (no welding yet).
pub fn concat(meshes: &[PageMesh]) -> PageMesh {
    let mut out = PageMesh::default();
    for m in meshes {
        let v_off = out.positions.len() as u32;
        out.positions.extend_from_slice(&m.positions);
        out.normals.extend_from_slice(&m.normals);
        out.materials.extend_from_slice(&m.materials);
        out.indices.extend(m.indices.iter().map(|&i| i + v_off));
    }
    out
}

/// Build a LOD0 page source from its PxP chunks (page coords).
pub fn build_lod0_page_source(
    page_x: i32,
    page_z: i32,
    cfg: &ClodPagesConfig,
    world: WorldBounds,
) -> Result<PageSource, ClodBuildError> {
    let p = cfg.page.chunks_per_page as i32;
    let s = cfg.page.chunk_size as i32;

    let mut chunks: Vec<PageMesh> = Vec::with_capacity((p * p) as usize);
    for dz in 0..p {
        for dx in 0..p {
            chunks.push(mesh_chunk(page_x * p + dx, page_z * p + dz, cfg, world));
        }
    }
    if chunks.len() != (p * p) as usize {
        return Err(ClodBuildError::PageIncomplete(format!(
            "expected {} chunks, got {}",
            p * p,
            chunks.len()
        )));
    }

    let merged = concat(&chunks);
    let (mesh, weld) = weld_vertices(&merged, cfg.simplify.weld_epsilon_cells)?;
    let footprint = PageFootprint {
        min_x: (page_x * p * s) as f32,
        min_z: (page_z * p * s) as f32,
        max_x: ((page_x + 1) * p * s) as f32,
        max_z: ((page_z + 1) * p * s) as f32,
    };
    Ok(PageSource { mesh, footprint, weld })
}
